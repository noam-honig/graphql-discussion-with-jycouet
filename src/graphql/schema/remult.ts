import { entities } from '$shared/_entities'
import { createSchema } from 'graphql-yoga'
import { fs } from 'houdini'
import { remultGraphql } from 'remult/graphql'

const { typeDefs, resolvers } = remultGraphql({ entities })

// Houdini needs this file to start! We can write a dummy one so that onstart there is something to real. Then it will be regenerated?
// TODO JYC
fs.writeFile('./src/graphql/schema.graphql', typeDefs)

//  remult GraphQL
export const remultSchema = createSchema({ typeDefs, resolvers })
