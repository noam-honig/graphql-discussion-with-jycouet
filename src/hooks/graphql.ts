import type { Handle } from '@sveltejs/kit';
import { remultGraphql } from '../remult_tmp/graphql';

import { createSchema, createYoga } from 'graphql-yoga';
import { remultApi } from './remult';
import { fs } from 'houdini';

const { typeDefs, resolvers } = remultGraphql(remultApi);

// Houdini needs this file to start! We can write a dummy one so that onstart there is something to real. Then it will be regenerated?
// TODO JYC
console.log('coucou');
fs.writeFile('./src/graphql/schema.graphql', typeDefs);

export const handleGraphql = (options?: { endpoint: string }): Handle => {
	const { endpoint } = {
		endpoint: '/api/graphql',
		...options
	};

	const yoga = createYoga({
		graphiql: {
			defaultQuery: `# Welcome to Remult GraphQL

# In this editor you can test your operations (query, mutation, subscription) 
# against the Remult GraphQL API.
# You can also discover the API using the built-in documentation explorer.
			
query My_First_Query {
	tasks(orderBy: { title: ASC }) {
		id
		title
	}
}
			`,
			title: 'KitQL - Remult'
		},
		graphqlEndpoint: endpoint,
		schema: createSchema({
			typeDefs,
			resolvers
		})
	});
	return async ({ event, resolve }) => {
		if (event.url && event.url.pathname === endpoint) {
			return await yoga.handleRequest(event.request, {});
		}
		return resolve(event);
	};
};
