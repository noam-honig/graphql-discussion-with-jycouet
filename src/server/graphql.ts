import { EntityMetadata, Field } from "remult";
import { RemultServerCore } from "remult/server";

type Field = {
  key: string;
  args?: string;
  value: string;
  comment?: string;
  order?: number;
};

type GraphQLType = {
  key: string;
  kind: "type_impl_node" | "type" | "input";
  fields: Field[];
  query: {
    orderBy: string[];
    whereType: string[];
    whereTypeSubFields: string[];
    resultProcessors: ((item: any) => void)[];
  };
  mutation: {
    create: {
      input?: GraphQLType;
      payload?: GraphQLType;
    };
    update: {
      input?: GraphQLType;
      payload?: GraphQLType;
    };
    delete: {
      payload?: GraphQLType;
    };
  };
  order?: number;
};

let _removeComments = false;
export function remultGraphql(
  api: RemultServerCore<any>,
  options?: { removeComments?: Boolean }
) {
  const { removeComments } = { removeComments: false, ...options };

  if (removeComments) {
    _removeComments = true;
  }

  let server = api["get internal server"]();
  const entities = server.getEntities();

  let types: GraphQLType[] = [];

  let root: Record<string, any> = {};
  let resolversQuery: Record<string, unknown> = {};
  let resolversMutation: Record<string, unknown> = {};
  let resolvers = { Query: resolversQuery, Mutation: resolversMutation };

  function upsertTypes(
    key: string,
    kind: "type_impl_node" | "type" | "input" = "type",
    order: number = 0
  ) {
    let t = types.find((t) => t.key === key);
    if (!t) {
      types.push(
        (t = {
          key,
          kind,
          fields: [],
          query: {
            orderBy: [],
            whereType: [],
            whereTypeSubFields: [],
            resultProcessors: [],
          },
          mutation: {
            create: {},
            update: {},
            delete: {},
          },
          order,
        })
      );
      if (kind === "type_impl_node") {
        t.fields.push({
          key: "nodeId",
          value: "ID!",
        });
      }
    }
    return t;
  }

  // Where - GraphQL primitives
  for (const whereType of ["String", "Int", "Float", "Boolean", "ID"]) {
    const currentWhere = upsertTypes(`Where${whereType}`, "input", 20);
    const currentWhereNullable = upsertTypes(
      `Where${whereType}Nullable`,
      "input",
      20
    );

    // For everyone
    for (const operator of ["eq", "ne", "in"]) {
      const field = {
        key: operator,
        value: operator === "in" ? `[${whereType}!]` : whereType,
      };
      currentWhere.fields.push(field);
      currentWhereNullable.fields.push(field);
    }

    // only for specific types
    if (
      whereType === "String" ||
      whereType === "Int" ||
      whereType === "Float"
    ) {
      for (const operator of ["gt", "gte", "lt", "lte"]) {
        const field = {
          key: operator,
          value: whereType,
        };
        currentWhere.fields.push(field);
        currentWhereNullable.fields.push(field);
      }
    }

    if (whereType === "String") {
      for (const operator of ["st"]) {
        const field = {
          key: operator,
          value: whereType,
        };
        currentWhere.fields.push(field);
        currentWhereNullable.fields.push(field);
      }
    }

    // add only for nullable
    currentWhereNullable.fields.push({
      key: "null",
      value: "Boolean",
    });
  }

  for (const meta of entities) {
    const orderByFields: string[] = [];

    let key = meta.key;

    const currentType = upsertTypes(getMetaType(meta), "type_impl_node");

    if (key) {
      const root_query = upsertTypes("Query", "type", -10);
      const queryArgsList =
        `limit: Int, page: Int, ` +
        `orderBy: ${key}OrderBy, ` +
        `where: ${key}Where`;

      root_query.fields.push({
        key: toPascalCase(getMetaType(meta)),
        args: `id: ID!`,
        value: `${getMetaType(meta)}`,
        comment: `Get \`${getMetaType(meta)}\` entity`,
      });
      // list
      root_query.fields.push({
        key,
        args: queryArgsList,
        value: `[${getMetaType(meta)}!]!`,
        comment: `List all \`${getMetaType(
          meta
        )}\` entity (with pagination, sorting and filtering)`,
      });
      resolversQuery[key] = (
        origItem: any,
        args: any,
        req: any,
        gqlInfo: any
      ) => root[key](args, req, gqlInfo);

      const root_mutation = upsertTypes("Mutation", "type", -9);

      // create
      const createInput = `Create${getMetaType(meta)}Input`;
      const createPayload = `Create${getMetaType(meta)}Payload`;
      const createResolverKey = `create${getMetaType(meta)}`;
      root_mutation.fields.push({
        key: createResolverKey,
        args: `input: ${createInput}!`,
        value: `${createPayload}`,
        comment: `Create a new \`${getMetaType(meta)}\``,
      });
      resolversMutation[createResolverKey] = (
        origItem: any,
        args: any,
        req: any,
        gqlInfo: any
      ) => root[createResolverKey](args, req, gqlInfo);
      currentType.mutation.create.input = upsertTypes(createInput, "input");

      currentType.mutation.create.payload = upsertTypes(createPayload);
      currentType.mutation.create.payload.fields.push({
        key: `${toPascalCase(getMetaType(meta))}`,
        value: `${getMetaType(meta)}`,
      });

      // update
      const updateInput = `Update${getMetaType(meta)}Input`;
      const updatePayload = `Update${getMetaType(meta)}Payload`;
      const updateResolverKey = `update${getMetaType(meta)}`;
      root_mutation.fields.push({
        key: updateResolverKey,
        args: `id: ID!, patch: ${updateInput}!`,
        value: `${updatePayload}`,
        comment: `Update a \`${getMetaType(meta)}\``,
      });

      currentType.mutation.update.input = upsertTypes(updateInput, "input");

      currentType.mutation.update.payload = upsertTypes(updatePayload);
      currentType.mutation.update.payload.fields.push({
        key: `${toPascalCase(getMetaType(meta))}`,
        value: `${getMetaType(meta)}`,
      });

      // delete
      const deletePayload = `Delete${getMetaType(meta)}Payload`;
      const deleteResolverKey = `delete${getMetaType(meta)}`;
      root_mutation.fields.push({
        key: deleteResolverKey,
        args: `id: ID!`,
        value: `${deletePayload}`,
        comment: `Delete a \`${getMetaType(meta)}\``,
      });

      currentType.mutation.delete.payload = upsertTypes(deletePayload);
      currentType.mutation.delete.payload.fields.push({
        key: `deleted${getMetaType(meta)}Id`,
        value: "ID",
      });

      const whereTypeFields: string[] = [];
      for (const f of meta.fields) {
        let whereFields: string[] = [];
        let type = "String";
        switch (f.valueType) {
          case Boolean:
            type = "Boolean";
            break;
          case Number:
            {
              if (
                f.valueConverter?.fieldTypeInDb === "integer" ||
                f.valueConverter?.fieldTypeInDb === "autoincrement"
              )
                type = "Int";
              else type = "Float";
            }
            break;
        }
        let ref = entities.find((i) => i.entityType === f.valueType);
        if (ref !== undefined) {
          // will do: Task.category
          currentType.fields.push({
            key: f.key,
            value: `${getMetaType(ref)}${f.allowNull ? "" : "!"}`,
            comment: f.caption,
          });
          const refKey = ref.key;
          currentType.query.resultProcessors.push((r) => {
            const val = r[f.key];
            if (val === null || val === undefined) return null;
            r[f.key] = async (args: any, req: any, gqlInfo: any) => {
              const queryResult: any[] = await root[refKey](
                {
                  ...args.where,
                  where: { id: val },
                  options: { limit: 1 },
                },
                req,
                gqlInfo
              );
              if (queryResult.length > 0) return queryResult[0];
              return null;
            };
          });

          // will do: Category.tasks
          let refT = upsertTypes(getMetaType(ref), "type_impl_node");
          refT.fields.push({
            key,
            args: queryArgsList,
            value: `[${getMetaType(meta)}!]!`,
            order: 10,
            comment: `List all \`${getMetaType(meta)}\` of \`${refKey}\``,
          });
          refT.query.resultProcessors.push((r) => {
            const val = r.id;
            r[key] = async (args: any, req: any, gqlInfo: any) => {
              return await root[key](
                {
                  where: { ...args.where, [f.key]: val },
                  options: { ...args.limit, ...args.page, ...args.orderBy },
                },
                req,
                gqlInfo
              );
            };
          });
        } else {
          currentType.fields.push({
            key: f.key,
            value: `${type}${f.allowNull ? "" : "!"}`,
            comment: f.caption,
          });
        }

        // sorting
        orderByFields.push(`${f.key}: OrderBydirection`);

        // helper
        const it_is_not_at_ref = ref === undefined;

        // where
        if (it_is_not_at_ref) {
          whereTypeFields.push(
            `${f.key}: Where${type}${f.allowNull ? "Nullable" : ""}`
          );
        }

        // Todo => Complete should not be a MUST. Where to look?
        // create
        if (!f.dbReadOnly && it_is_not_at_ref) {
          currentType.mutation.create.input.fields.push({
            key: f.key,
            value: `${type}${f.allowNull ? "" : "!"}`,
          });
        }

        // update
        if (!f.dbReadOnly && it_is_not_at_ref) {
          currentType.mutation.update.input.fields.push({
            key: f.key,
            value: `${type}${f.allowNull ? "" : "!"}`,
          });
        }
      }

      currentType.query.orderBy.push(
        blockFormat({
          prefix: `input ${key}OrderBy`,
          data: orderByFields,
          comment: `OrderBy options for \`${key}\``,
        })
      );

      whereTypeFields.push(`OR: [${key}Where!]`);
      currentType.query.whereType.push(
        blockFormat({
          prefix: `input ${key}Where`,
          data: whereTypeFields,
          comment: `Where options for \`${key}\``,
        })
      );

      root[key] = async (arg1: any, req: any) => {
        const { limit, page, orderBy, where } = arg1;

        return new Promise(async (res, error) => {
          server.run(req, async () => {
            let dApi = await server.getDataApi(req, meta);
            let result: any;
            let err: any;
            await dApi.getArray(
              {
                success: (x) => {
                  return (result = x.map((y: any) => {
                    currentType.query.resultProcessors.forEach((z) => z(y));
                    return y;
                  }));
                },
                created: () => {},
                deleted: () => {},
                error: (x) => (err = x),
                forbidden: () => (err = "forbidden"),
                notFound: () => (err = "not found"),
                progress: () => {},
              },
              {
                get: (key) => {
                  if (limit && key === "_limit") {
                    return limit;
                  }
                  if (page && key === "_page") {
                    return page;
                  }
                  if (orderBy) {
                    if (key === "_sort") {
                      const sort_keys: string[] = [];
                      Object.keys(orderBy).forEach((sort_key) => {
                        sort_keys.push(sort_key);
                      });
                      if (sort_keys.length > 0) {
                        return sort_keys.join(",");
                      }
                    } else if (key === "_order") {
                      const sort_directions: string[] = [];
                      Object.keys(orderBy).forEach((sort_key) => {
                        const direction = orderBy[sort_key].toLowerCase();
                        sort_directions.push(direction);
                      });
                      if (sort_directions.length > 0) {
                        return sort_directions.join(",");
                      }
                    }
                  }
                  if (where) {
                    // TODO: OR management?

                    const whereAND: string[] = [];
                    Object.keys(where).forEach((w) => {
                      const subWhere = where[w];
                      Object.keys(subWhere).forEach((sw) => {
                        let map = `${w}.${sw}`;

                        if (map.endsWith(".eq")) {
                          map = `${w}`;
                        }

                        if (map === key) {
                          whereAND.push(subWhere[sw]);
                          return subWhere[sw];
                        }
                      });
                    });
                    if (whereAND.length > 0) {
                      return whereAND.join(",");
                    }
                  }
                },
              }
            );
            if (err) {
              error(err);
              return;
            }
            res(result);
          });
        });
      };

      root[createResolverKey] = async (arg1: any, req: any) => {
        console.log(`createResolverKey`, arg1);
      };

      root[updateResolverKey] = async (arg1: any, req: any) => {
        console.log(`updateResolverKey`, arg1);
      };

      root[deleteResolverKey] = async (arg1: any, req: any) => {
        console.log(`deleteResolverKey`, arg1);
      };
    }
  }

  return {
    resolvers,
    rootValue: root,
    schema: `${types
      .sort((a, b) => (a.order ? a.order : 0) - (b.order ? b.order : 0))
      .map(({ key, kind, fields, query }) => {
        const { orderBy, whereType, whereTypeSubFields } = query;

        let prefix = `${kind} ${key}`;
        if (kind === "type_impl_node") {
          prefix = `type ${key} implements Node`;
        }

        const type = blockFormat({
          prefix,
          data: fields
            .sort((a, b) => (a.order ? a.order : 0) - (b.order ? b.order : 0))
            .map((field) => fieldFormat(field)),
          comment:
            key === "Query"
              ? `Represents all Remult entities.`
              : key === "Mutation"
              ? `Represents all Remult Mutations available on Remult.`
              : `Represents \`${key}\` entity.`,
        });

        const orderByStr =
          orderBy.length > 0 ? `\n\n${orderBy.join("\n\n")}` : ``;
        const whereTypeStr =
          whereType.length > 0 ? `\n\n${whereType.join("\n\n")}` : ``;
        const whereTypeSubFieldsStr =
          whereTypeSubFields.length > 0
            ? `\n\n${whereTypeSubFields.join("\n\n")}`
            : ``;
        return `${type}${orderByStr}${whereTypeStr}${whereTypeSubFieldsStr}`;
      })
      .join(`\n\n`)}

${schemaGlobal}
`,
  };
}

const schemaGlobal = `"""
Determines the order of items returned
"""
enum OrderBydirection {
  """
  Sort data in ascending order
  """
  ASC
  """
  Sort data in descending order
  """
  DESC
}

"""
Node interface of remult entities
"""
interface Node {
  nodeId: ID!
}`;

function blockFormat(obj: { prefix: string; data: string[]; comment: string }) {
  if (obj.data.length === 0) {
    return ``;
  }

  const str = `${obj.prefix} {
  ${obj.data.join("\n  ")}
}`;

  let commentsStr = `"""
${obj.comment}
"""
`;

  if (_removeComments) {
    commentsStr = ``;
  }

  return `${commentsStr}${str}`;
}

function fieldFormat(field: Field) {
  const key_value = `${field.key}${field.args ? ` (${field.args})` : ``}: ${
    field.value
  }`;

  if (!_removeComments && field.comment) {
    return `"""
  ${field.comment}
  """
  ${key_value}`;
  }

  return key_value;
}

function getMetaType(entityMeta: EntityMetadata) {
  return entityMeta.entityType.name;
}

function toPascalCase(str: string) {
  return str
    .split("")
    .map((c, i) => (i === 0 ? c.toLowerCase() : c))
    .join("");
}
