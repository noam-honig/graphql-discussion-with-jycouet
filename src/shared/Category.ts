import { Entity, Fields, FieldType } from 'remult'

export
@FieldType({ displayValue: (_, v) => v?.name })
@Entity('categories', { allowApiCrud: true })
class Category {
  @Fields.string({
    allowApiUpdate: false,
    saving: async (_, ref) => {
      // created a consistent id for testing
      ref.value = (await ref.entityRef.repository.count()).toString()
    },
  })
  id = ''
  @Fields.string({
    withLink: true,
  })
  name = ''
}
