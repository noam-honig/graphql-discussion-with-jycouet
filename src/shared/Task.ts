import { Entity, Field, Fields } from 'remult'

import { Category } from './Category'

export
@Entity('tasks', {
  allowApiCrud: true,
  // We will set this later to see the behavior in GraphQL
  // allowApiInsert: 'admin',
})
class Task {
  // @Fields.cuid()
  @Fields.autoIncrement()
  id = 0
  @Fields.string({
    caption: 'The Title',
    validate: task => {
      if (task.title.length < 3) throw Error('Too short')
    },
  })
  title = ''
  @Fields.boolean({ caption: 'Is it completed?' })
  completed = false
  @Fields.object({ dbName: 'the_priority' })
  thePriority = Priority.Low
  @Field(() => Category, { allowNull: true })
  category?: Category
}

export enum Priority {
  Low,
  High,
  Critical,
}
