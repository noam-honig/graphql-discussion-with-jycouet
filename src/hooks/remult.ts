import { remultSveltekit } from 'remult/remult-sveltekit';
import { Task } from '../shared/Task';
import { Category } from '../shared/Category';

export const remultApi = remultSveltekit({
	entities: [Task, Category]
});
