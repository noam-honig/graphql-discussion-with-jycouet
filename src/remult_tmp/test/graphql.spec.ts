import { createSchema, createYoga } from 'graphql-yoga'
import { InMemoryDataProvider, remult } from 'remult'
import { remultExpress, type RemultExpressServer } from 'remult/remult-express'
import { beforeEach, describe, expect, it } from 'vitest'

import { Category } from '../../shared/Category'
import { Task } from '../../shared/Task'
import { remultGraphql } from '../graphql'

// import { remultSveltekit, type RemultSveltekitServer } from 'remult/remult-sveltekit';

describe('graphql-connection', () => {
  let api: RemultExpressServer
  async function withRemult<T>(what: () => Promise<T>): Promise<T> {
    return await api['get internal server']().run({} as any, what)
  }
  let gql: (gql: string) => Promise<any>
  beforeEach(async () => {
    // api = remultSveltekit({
    api = remultExpress({
      logApiEndPoints: false, // We don't need this in tests
      dataProvider: new InMemoryDataProvider(),
      entities: [Task, Category],
    })
    const { typeDefs, resolvers } = remultGraphql(api)

    const yoga = createYoga({
      schema: createSchema({
        typeDefs,
        resolvers,
      }),
    })
    gql = async (query: string) => {
      return await yoga.getResultForParams({
        request: {} as any,
        params: {
          query,
        },
      })
    }
  })
  it('test mutation create', async () => {
    const result = await gql(`
mutation {
  createTask(input: {title: "testing"}) {
    task {
      id
      title
    }
  }
}`)
    expect(await withRemult(() => remult.repo(Task).count())).toBe(1)
  })
  it('test mutation update', async () => {
    await withRemult(() => remult.repo(Task).insert({ title: 'aaa' }))
    console.log("HAHAHA")
    const result = await gql(`
mutation {
  updateTask(id:1, patch: {title: "bbb"}) {
    task {
      id
      title
    }
  }
}`)
 //   expect(result).toBe('')
 //   expect((await withRemult(() => remult.repo(Task).findFirst())).title).toBe('b')
  })
  it('test graphql', async () => {
    await withRemult(async () => {
      await remult.repo(Task).insert([{ title: 'task c' }])
      await remult.repo(Task).insert([{ title: 'task b' }])
      await remult.repo(Task).insert([{ title: 'task a' }])
      expect(await remult.repo(Task).count()).toBe(3)
    })

    const result = await gql(/* GraphQL */ `
      query Tasks {
        tasks(orderBy: { title: ASC }) {
          id
          title
          completed
        }
      }
    `)

    expect(result).toMatchInlineSnapshot(`
			{
			  "data": {
			    "tasks": [
			      {
			        "completed": false,
			        "id": 3,
			        "title": "task a",
			      },
			      {
			        "completed": false,
			        "id": 2,
			        "title": "task b",
			      },
			      {
			        "completed": false,
			        "id": 1,
			        "title": "task c",
			      },
			    ],
			  },
			}
		`)
  })
  it('test basics', async () => {
    // rmv removeComments is very handy for testing!
    const { typeDefs } = remultGraphql(api, {
      removeComments: true,
    })

    expect(typeDefs).toMatchInlineSnapshot(`
      "type Query {
          task(id: ID!): Task
          tasks(limit: Int, page: Int, orderBy: tasksOrderBy, where: tasksWhere): [Task!]!
          tasksConnection(first: Int, after: String, last: Int, before: String, orderBy: tasksOrderBy, where: tasksWhere): TaskConnection
          category(id: ID!): Category
          categories(limit: Int, page: Int, orderBy: categoriesOrderBy, where: categoriesWhere): [Category!]!
          categoriesConnection(first: Int, after: String, last: Int, before: String, orderBy: categoriesOrderBy, where: categoriesWhere): CategoryConnection
          node(nodeId: ID!): Node
      }

      type Mutation {
          createTask(input: CreateTaskInput!): CreateTaskPayload
          updateTask(id: ID!, patch: UpdateTaskInput!): UpdateTaskPayload
          deleteTask(id: ID!): DeleteTaskPayload
          createCategory(input: CreateCategoryInput!): CreateCategoryPayload
          updateCategory(id: ID!, patch: UpdateCategoryInput!): UpdateCategoryPayload
          deleteCategory(id: ID!): DeleteCategoryPayload
      }

      type Task implements Node {
          id: Int!
          title: String!
          completed: Boolean!
          category: Category
          nodeId: ID!
      }

      input tasksOrderBy {
        id: OrderByDirection
        title: OrderByDirection
        completed: OrderByDirection
        category: OrderByDirection
      }

      input tasksWhere {
        id: WhereInt
        title: WhereString
        completed: WhereBoolean
        OR: [tasksWhere!]
        AND: [tasksWhere!]
      }

      type TaskConnection {
          totalCount: Int!
          edges: [TaskEdge!]!
          pageInfo: PageInfo!
      }

      type TaskEdge {
          node: Task!
          cursor: String!
      }

      input CreateTaskInput {
          title: String
          completed: Boolean
      }

      type CreateTaskPayload {
          task: Task
      }

      input UpdateTaskInput {
          title: String
          completed: Boolean
      }

      type UpdateTaskPayload {
          task: Task
      }

      type DeleteTaskPayload {
          deletedTaskId: ID
      }

      type Category implements Node {
          id: String!
          name: String!
          tasks(limit: Int, page: Int, orderBy: tasksOrderBy, where: tasksWhere): [Task!]!
          nodeId: ID!
      }

      input categoriesOrderBy {
        id: OrderByDirection
        name: OrderByDirection
      }

      input categoriesWhere {
        id: WhereString
        name: WhereString
        OR: [categoriesWhere!]
        AND: [categoriesWhere!]
      }

      type CategoryConnection {
          totalCount: Int!
          edges: [CategoryEdge!]!
          pageInfo: PageInfo!
      }

      type CategoryEdge {
          node: Category!
          cursor: String!
      }

      input CreateCategoryInput {
          id: String
          name: String
      }

      type CreateCategoryPayload {
          category: Category
      }

      input UpdateCategoryInput {
          id: String
          name: String
      }

      type UpdateCategoryPayload {
          category: Category
      }

      type DeleteCategoryPayload {
          deletedCategoryId: ID
      }

      input WhereString {
          eq: String
          ne: String
          in: [String!]
          gt: String
          gte: String
          lt: String
          lte: String
          st: String
      }

      input WhereStringNullable {
          eq: String
          ne: String
          in: [String!]
          gt: String
          gte: String
          lt: String
          lte: String
          st: String
          null: Boolean
      }

      input WhereInt {
          eq: Int
          ne: Int
          in: [Int!]
          gt: Int
          gte: Int
          lt: Int
          lte: Int
      }

      input WhereIntNullable {
          eq: Int
          ne: Int
          in: [Int!]
          gt: Int
          gte: Int
          lt: Int
          lte: Int
          null: Boolean
      }

      input WhereFloat {
          eq: Float
          ne: Float
          in: [Float!]
          gt: Float
          gte: Float
          lt: Float
          lte: Float
      }

      input WhereFloatNullable {
          eq: Float
          ne: Float
          in: [Float!]
          gt: Float
          gte: Float
          lt: Float
          lte: Float
          null: Boolean
      }

      input WhereBoolean {
          eq: Boolean
          ne: Boolean
          in: [Boolean!]
      }

      input WhereBooleanNullable {
          eq: Boolean
          ne: Boolean
          in: [Boolean!]
          null: Boolean
      }

      input WhereID {
          eq: ID
          ne: ID
          in: [ID!]
      }

      input WhereIDNullable {
          eq: ID
          ne: ID
          in: [ID!]
          null: Boolean
      }

      type PageInfo {
          endCursor: String!
          hasNextPage: Boolean!
          hasPreviousPage: Boolean!
          startCursor: String!
      }

      enum OrderByDirection {
          ASC
          DESC
      }

      interface Node {
          nodeId: ID!
      }
      "
    `)
  })

  it('test get values', async () => {
    withRemult(async () => {
      await remult.repo(Task).insert([{ title: 'task a' }])
      expect(await remult.repo(Task).count()).toBe(1)
    })
    const { resolvers } = remultGraphql(api)
    expect((await (resolvers.Query.tasks as any)(undefined, {}, {})).length).toBe(1)
  })
})
