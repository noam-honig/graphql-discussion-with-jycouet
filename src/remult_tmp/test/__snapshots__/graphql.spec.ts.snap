// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`graphql-connection > gets related entities 1`] = `
{
  "data": {
    "tasks": {
      "items": [
        {
          "category": {
            "name": "c1",
            "tasks": {
              "items": [
                {
                  "title": "task a",
                },
              ],
            },
          },
          "title": "task a",
        },
        {
          "category": {
            "name": "c2",
            "tasks": {
              "items": [
                {
                  "title": "task b",
                },
              ],
            },
          },
          "title": "task b",
        },
      ],
    },
  },
}
`;

exports[`graphql-connection > test basics 1`] = `
"type Query {
    task(id: ID!): Task
    tasks(limit: Int, page: Int, orderBy: tasksOrderBy, where: tasksWhere): TaskConnection
    category(id: ID!): Category
    categories(limit: Int, page: Int, orderBy: categoriesOrderBy, where: categoriesWhere): CategoryConnection
    node(nodeId: ID!): Node
}

type Mutation {
    createTask(input: CreateTaskInput!, clientMutationId: String): CreateTaskPayload
    updateTask(id: ID!, patch: UpdateTaskInput!, clientMutationId: String): UpdateTaskPayload
    deleteTask(id: ID!, clientMutationId: String): DeleteTaskPayload
    createCategory(input: CreateCategoryInput!, clientMutationId: String): CreateCategoryPayload
    updateCategory(id: ID!, patch: UpdateCategoryInput!, clientMutationId: String): UpdateCategoryPayload
    deleteCategory(id: ID!, clientMutationId: String): DeleteCategoryPayload
}

type Task implements Node {
    id: Int!
    title: String!
    completed: Boolean!
    thePriority: String!
    category: Category
    nodeId: ID!
}

input tasksOrderBy {
  id: OrderByDirection
  title: OrderByDirection
  completed: OrderByDirection
  thePriority: OrderByDirection
  category: OrderByDirection
}

input tasksWhere {
  id: WhereInt
  title: WhereString
  completed: WhereBoolean
  thePriority: WhereString
  OR: [tasksWhere!]
  AND: [tasksWhere!]
}

type TaskConnection {
    totalCount: Int!
    items: [Task!]!
}

input CreateTaskInput {
    title: String
    completed: Boolean
    thePriority: String
    category: ID
}

type CreateTaskPayload {
    task: Task
    clientMutationId: String
}

input UpdateTaskInput {
    title: String
    completed: Boolean
    thePriority: String
    category: ID
}

type UpdateTaskPayload {
    task: Task
    clientMutationId: String
}

type DeleteTaskPayload {
    id: ID
    clientMutationId: String
}

type Category implements Node {
    id: String!
    name: String!
    tasks(limit: Int, page: Int, orderBy: tasksOrderBy, where: tasksWhere): TaskConnection
    nodeId: ID!
}

input categoriesOrderBy {
  id: OrderByDirection
  name: OrderByDirection
}

input categoriesWhere {
  id: WhereString
  name: WhereString
  OR: [categoriesWhere!]
  AND: [categoriesWhere!]
}

type CategoryConnection {
    totalCount: Int!
    items: [Category!]!
}

input CreateCategoryInput {
    name: String
}

type CreateCategoryPayload {
    category: Category
    clientMutationId: String
}

input UpdateCategoryInput {
    name: String
}

type UpdateCategoryPayload {
    category: Category
    clientMutationId: String
}

type DeleteCategoryPayload {
    id: ID
    clientMutationId: String
}

input WhereString {
    eq: String
    ne: String
    in: [String!]
    nin: [String!]
    gt: String
    gte: String
    lt: String
    lte: String
    st: String
}

input WhereStringNullable {
    eq: String
    ne: String
    in: [String!]
    nin: [String!]
    gt: String
    gte: String
    lt: String
    lte: String
    st: String
    null: Boolean
}

input WhereInt {
    eq: Int
    ne: Int
    in: [Int!]
    nin: [Int!]
    gt: Int
    gte: Int
    lt: Int
    lte: Int
}

input WhereIntNullable {
    eq: Int
    ne: Int
    in: [Int!]
    nin: [Int!]
    gt: Int
    gte: Int
    lt: Int
    lte: Int
    null: Boolean
}

input WhereFloat {
    eq: Float
    ne: Float
    in: [Float!]
    nin: [Float!]
    gt: Float
    gte: Float
    lt: Float
    lte: Float
}

input WhereFloatNullable {
    eq: Float
    ne: Float
    in: [Float!]
    nin: [Float!]
    gt: Float
    gte: Float
    lt: Float
    lte: Float
    null: Boolean
}

input WhereBoolean {
    eq: Boolean
    ne: Boolean
    in: [Boolean!]
    nin: [Boolean!]
}

input WhereBooleanNullable {
    eq: Boolean
    ne: Boolean
    in: [Boolean!]
    nin: [Boolean!]
    null: Boolean
}

input WhereID {
    eq: ID
    ne: ID
    in: [ID!]
    nin: [ID!]
}

input WhereIDNullable {
    eq: ID
    ne: ID
    in: [ID!]
    nin: [ID!]
    null: Boolean
}

enum OrderByDirection {
    ASC
    DESC
}

interface Node {
    nodeId: ID!
}
"
`;

exports[`graphql-connection > test get single task by id 1`] = `
{
  "data": {
    "task": {
      "id": 2,
      "title": "bbb",
    },
  },
}
`;

exports[`graphql-connection > test graphql 1`] = `
{
  "data": {
    "tasks": {
      "items": [
        {
          "completed": false,
          "id": 3,
          "title": "task a",
        },
        {
          "completed": false,
          "id": 2,
          "title": "task b",
        },
        {
          "completed": false,
          "id": 1,
          "title": "task c",
        },
      ],
    },
  },
}
`;

exports[`graphql-connection > test mutation create 1`] = `
{
  "data": {
    "createTask": {
      "task": {
        "id": 1,
        "title": "testing",
      },
    },
  },
}
`;

exports[`graphql-connection > test mutation create 2`] = `
[
  Task {
    "category": null,
    "completed": false,
    "id": 1,
    "thePriority": 1,
    "title": "testing",
  },
]
`;

exports[`graphql-connection > test mutation delete 1`] = `
{
  "data": {
    "deleteTask": {
      "id": "2",
    },
  },
}
`;

exports[`graphql-connection > test mutation delete 2`] = `
[
  Task {
    "category": null,
    "completed": false,
    "id": 1,
    "thePriority": 1,
    "title": "task a",
  },
  Task {
    "category": null,
    "completed": false,
    "id": 3,
    "thePriority": 1,
    "title": "task c",
  },
]
`;

exports[`graphql-connection > test mutation update 1`] = `
{
  "data": {
    "updateTask": {
      "task": {
        "id": 1,
        "title": "bbb",
      },
    },
  },
}
`;
